### PR TITLE
Feature/interlok 4451

### PR DESCRIPTION
**Pull request by telus-terchu**
## Motivation

https://adaptris.atlassian.net/browse/INTERLOK-4451

## Modification

- Added AddMetadataExpressionService
- Added AddMetadataExpressionServiceTest
- modified AddMetadataService to allow subclasses to override some behaviour

## PR Checklist

- [x] been self-reviewed.
- [x ] Added javadocs for most classes and all non-trivial methods
- [x] Added supporting annotations (like @XStreamAlias / @ComponentProfile)
- [ ] Added DefaultValue annotation when there is a default value (e.g. @DefaultValue('true'))
- [ ] Added validation annotation (@NotNull...) when required and add @Valid when nested class has some validation
- [ ] Checked that @NotNull and @NotBlank annotations have a meaningful message when appropriate.
- [ ] Checked that new 3rd party dependencies are appropriately licensed
- [x] Added comments explaining the "why" and the intent of the code wherever it would not be obvious for an unfamiliar reader
- [x] Added unit tests or modified existing tests to cover new code paths
- [ ] Tested new/updated components in the UI and at runtime in an Interlok instance
- [ ] Reviewed java class members so that missing annotations are added (InputFieldDefault/ComponentProfile etc)
- [ ] Checked that javadoc generation doesn't report errors
- [ ] Checked the display of the component in the UI
- [ ] Remove any config/license annotations
- [ ] Check the gradle build file to make sure the dependencies section is more explicit "implementation/api".

## Result

Users can use this service to resolve metadata keys/values:

```
<add-metadata-expression-service>
  <key>%message{exisiting-key}</key>
  <value>%message{valueA}, %message{valueB}, %message{valueC}</value>
</add-metadata-expression-service>
```

## Testing

- run AddMetadataExpressionServiceTest
- test with an adapter

### CODE REVIEW

File: interlok-core/src/test/java/com/adaptris/core/services/metadata/MetadataAppenderServiceTest.java
Comment (sebastien-belin-adp) [inline interlok-core/src/test/java/com/adaptris/core/services/metadata/MetadataAppenderServiceTest.java:1]: @telus-terchu Why did you remove this test class?

**Review by sebastien-belin-adp — APPROVED** [review]

File: interlok-core/src/test/java/com/adaptris/core/services/metadata/MetadataAppenderServiceTest.java
Comment (telus-terchu) [reply interlok-core/src/test/java/com/adaptris/core/services/metadata/MetadataAppenderServiceTest.java:1]: @sebastien-belin-adp thanks for the pick up I'm not sure why I deleted it, but I've reinstated it now
